Codeinwp/optimole-wp
# Rollback/Offloading Improvements

### All Submissions:

 

### Changes proposed in this Pull Request:
This PR aims to improve the offloading/rollback process in Optimole. Based on tests run on different servers, the offloading is about 60% faster on average based on tests done on websites with 300-500 posts (images being about 3 times the posts).

I tried to also run tests on sites to compare with 1000+ posts, but it took pretty long to run head-to-head tests between the older method & newer method. Overall, things were progressing faster with the PR compared to older code.

Also includes:

- Improvements to how we handle single/selected image uploads and make it quicker instead of taking as much time as offloading all images.
- The process not runs using Cron so the user can leave the page, which makes it easier for bigger websites with thousands of posts to leave the page.
- It also has a Stop button if the user stops it.
- It also removes the estimated time, as it was not even close to being accurate.
- Take all the logic to server side to avoid thing being terribly slower if user's internet connection is weak.
- When process is ongoing and user is on the Optimole Dashboard, we send a request each 10 seconds to make sure Cron is triggered + also to update the progress bar.

I had also tried to run few consecutive cron jobs but then WP kept missing crons as some processes were faster than other, and number of cron jobs to run kept growing.

Closes https://github.com/Codeinwp/optimole-service/issues/851.

### How to test the changes in this Pull Request:

1. Make sure offloading and rollback work properly and are significantly quicker than previously.
2. Make sure single image offloading/rollback also works properly and don't take more than a couple of minutes max.
3. Make sure the website doesn't become too slow and non-accessible.
4. Make sure you can leave the page and come back to it, and it should still show the offloading/rollback ongoing.
5. It is a bit complex so would be great if you can enable debugging and keep a log of your debug.log file, and post it while reporting issues.

### Other information:
